krun/platform.py

```diff
@@ -1140,19 +1153,45 @@ def _check_cpu_scaler(self):
                     fatal("The kernel is using '%s' for CPU scaling instead "
                           "of using 'acpi-cpufreq'" % v)
 
-        # Check "turbo boost" is disabled
-        # It really should be, as turbo boost is only available using pstates,
-        # and the code above is ensuring we are not. Let's check anyway.
+        # On some systems, you cannot turn turbo boost off in the BIOS, so we
+        # forcibly turn off turbo mode via the IA32_MISC_ENABLE MSR for cores
+        # which report as "supporting" turbo mode.
         debug("Checking 'turbo boost' is disabled")
-        if os.path.exists(LinuxPlatform.TURBO_DISABLED):
-            with open(LinuxPlatform.TURBO_DISABLED) as fh:
-                v = int(fh.read().strip())
-
-            if v != 1:
-                fatal("Machine has 'turbo boost' enabled. "
-                      "This should not happen, as this feature only applies to "
-                      "pstate CPU scaling and Krun just determined that "
-                      "the system is not!")
+
+        # If this file exists, then the pstate driver is loaded. It shouldn't be!
+        assert not os.path.exists(LinuxPlatform.TURBO_DISABLED)
+
+        # Query each core asking if it supports turbo mode. If it says "yes",
+        # then we turn it off. Depending on the CPU, the turbo settings may be
+        # shared across more than a single core. To be safe, we assume there is
+        # no sharing. After turning off turbo mode on all cores.
+        # XXX move loading msr
+        args = self.change_user_args("root") + ["modprobe", "msr"]
+        run_shell_cmd(" ".join(args))
+        query_program_path = os.path.join(UTILS_DIR, "query_turbo")
+        for core in xrange(self.num_cpus):
+            args = ["taskset", "-c", str(core), query_program_path]
```

Comment:
Like this.

Reply:
I see.

Reply:
Pretty nasty, I know, but it's the only way from userspace :(